src/CurveCallee.sol

```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published
+// by the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.6.12;
+pragma experimental ABIEncoderV2;
+
+interface GemJoinLike {
+    function dec() external view returns (uint256);
+    function gem() external view returns (address);
+    function exit(address, uint256) external;
+}
+
+interface DaiJoinLike {
+    function dai() external view returns (TokenLike);
+    function join(address, uint256) external;
+}
+
+interface TokenLike {
+    function approve(address, uint256) external;
+    function transfer(address, uint256) external;
+    function balanceOf(address) external view returns (uint256);
+    function symbol() external view returns (string memory);
+}
+
+interface CharterManagerLike {
+    function exit(address crop, address usr, uint256 val) external;
+}
+
+interface WstEthLike is TokenLike {
+    function unwrap(uint256 _wstEthAmount) external returns (uint256);
+    function stETH() external view returns (address);
+}
+
+interface CurveLike {
+    function exchange(int128 i, int128 j, uint256 dx, uint256 min_dy)
+        external returns (uint256 dy);
+}
+
+interface WethLike is TokenLike {
+    function deposit() external payable;
+}
+
+interface UniV3Like {
+    
+    struct Params {
+        bytes   path;
+        address recipient;
+        uint256 deadline;
+        uint256 amountIn;
+        uint256 amountOutMinimum;
+    }
+
+    function WETH9() external view returns (address);
+
+    function exactInput(UniV3Like.Params calldata params)
+        external payable returns (uint256 amountOut);
+}
+
+contract CurveCallee {
+    CurveLike               public curve;
+    UniV3Like               public uniV3;
+    DaiJoinLike             public daiJoin;
+    TokenLike               public dai;
+
+    uint256                 public constant RAY = 10 ** 27;
+
+    function add(uint x, uint y) internal pure returns (uint z) {
+        require((z = x + y) >= x, "ds-math-add-overflow");
+    }
+    function sub(uint x, uint y) internal pure returns (uint z) {
+        require((z = x - y) <= x, "ds-math-sub-underflow");
+    }
+    function divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = add(x, sub(y, 1)) / y;
+    }
```

Comment:
I am not getting any shadowing warning from this file, probably because no inline assembly is being used, but I still think it is a good practice. I will soon make a PR for this change in the vat so we get rid of all these warnings.

Reply:
done.